lib/apiservers/engine/backends/portmap/portmap.go

```diff
@@ -117,52 +143,68 @@ func (p *portMapper) forward(action iptables.Action, ip net.IP, port int, proto,
 		// value" by both iptables and ip6tables.
 		daddr = "0/0"
 	}
-	args := []string{"-t", string(iptables.Nat), string(action), "VIC",
-		"-i", srcIface,
-		"-p", proto,
-		"-d", daddr,
-		"--dport", strconv.Itoa(port),
-		"-j", "DNAT",
-		"--to-destination", net.JoinHostPort(destAddr, strconv.Itoa(destPort))}
-	if output, err := iptables.Raw(args...); err != nil {
-		return err
-	} else if len(output) != 0 {
-		return iptables.ChainError{Chain: "FORWARD", Output: output}
-	}
-
 	ipStr := ""
 	if ip != nil && !ip.IsUnspecified() {
 		ipStr = ip.String()
 	}
 
+	key := bindKey{ipStr, port}
 	switch action {
-	case iptables.Append:
-		p.bindings[bindKey{ipStr, port}] = nil
-
 	case iptables.Delete:
-		delete(p.bindings, bindKey{ipStr, port})
-	}
+		val := p.bindings[key] // lookup commands to reverse
+		if args, ok := val.([][]string); ok {
+			if err := iptablesDelete(args); err != nil {
+				return err
+			}
+			delete(p.bindings, bindKey{ipStr, port})
+			return nil
+		} else {
+			return fmt.Errorf("Failed to find unmap data for %s:%d", ipStr, port)
+		}
 
-	if output, err := iptables.Raw("-t", string(iptables.Filter), string(action), "VIC",
-		"-i", srcIface,
-		"-o", destIface,
-		"-p", proto,
-		"-d", destAddr,
-		"--dport", strconv.Itoa(destPort),
-		"-j", "ACCEPT"); err != nil {
-		return err
-	} else if len(output) != 0 {
-		return iptables.ChainError{Chain: "FORWARD", Output: output}
-	}
+	case iptables.Append:
+		var savedArgs [][]string
+
+		args := []string{"-t", string(iptables.Nat), string(action), "VIC",
+			"-i", srcIface,
+			"-p", proto,
+			"-d", daddr,
+			"--dport", strconv.Itoa(port),
+			"-j", "DNAT",
+			"--to-destination", net.JoinHostPort(destAddr, strconv.Itoa(destPort))}
+		if err := iptablesRunAndCheck(args); err != nil {
+			return err
+		}
+		savedArgs = append(savedArgs, args)
+		p.bindings[key] = savedArgs
```

Comment:
You can leave off the action arg (move it to the beginning of the slice so you can hack it off), and append it back on later when running delete.